### PR TITLE
chore(release): allow canary release

### PR DESCRIPTION
**Pull request by samouss**
This PR allows to release a "canary" version of the package. 

The version follows this format: `0.0.0-SHORT_SHA-1` where `SHORT_SHA-1` is the hash of the latest commit. The version is then published on npm with the tag `canary`. The command does not update the CHANGELOG or create a Git Tag. The usage of the version is mostly internal. Its purpose is to publish a non-stable version that can be used directly from other projects via npm (useful for the CI).

At some point, we might want to port the process on other projects as well. It would allow us to test packages directly from E2E tests or even on example projects that we maintain. I didn't use it (yet) to publish a version (I did test it though with the `pack` command). I'll do it once the PR is approved and merged.

Screenshot 2019-05-17 at 14 30 40

### CODE REVIEW

File: scripts/release.js [hunk 1]
```diff
@@ -20,29 +22,57 @@ const packageJson = require('../package.json');
...
   });
-});
+}
 
-function checkEnvironment() {
+function releaseStableVersion() {
+  shell.echo(`Algoliasearch-Helper release STABLE version`);
+
+  checkDevelopBranch();
+  checkCleanWorkdir();
```
Comment (samouss) [inline scripts/release.js]: Those two functions are the equivalent of the previous call `checkEnvironment()`.

File: scripts/release.js [hunk 2]
```diff
@@ -136,3 +170,16 @@ function goBackToDevelop() {
...
+function updateCanaryVersion(cb) {
+  const lastCommitHash = shell.exec('git rev-parse --short HEAD', {silent: true}).toString().trim();
+  const canaryVersion = `0.0.0-${lastCommitHash}`;
+
+  bumpVersion(canaryVersion, () => {
+    cb(canaryVersion);
+  });
+}
+
+function revertStableVersion(version, cb) {
```
Comment (samouss) [inline scripts/release.js]: I've used a function rather than inlining the call to be more explicit.

File: scripts/release.js [hunk 3]
```diff
@@ -20,29 +22,57 @@ const packageJson = require('../package.json');
...
+  releaseStableVersion();
+}
 
-checkEnvironment();
-mergeDevIntoMaster();
-showChangelog(shell);
-promptVersion(packageJson.version, (version) => {
-  bumpVersion(version, () => {
+function releaseCanaryVersion() {
+  shell.echo(`Algoliasearch-Helper release CANARY version`);
```
Comment (francoischalifour) [inline scripts/release.js]:
```suggestion
  shell.echo(`AlgoliaSearch Helper release CANARY version`);
```

File: scripts/release.js [hunk 4]
```diff
@@ -20,29 +22,57 @@ const packageJson = require('../package.json');
...
+  checkCleanWorkdir();
+  updateCanaryVersion(canaryVersion => {
     build();
-    updateChangelog(shell);
-    commitNewFiles(version);
-    publish();
-    goBackToDevelop();
+    publishOnNpm('canary');
+    revertStableVersion(packageJson.version, () => {
+      shell.echo(`Algoliasearch-Helper v${canaryVersion} released`);
```
Comment (francoischalifour) [inline scripts/release.js]:
```suggestion
      shell.echo(`AlgoliaSearch Helper v${canaryVersion} released`);
```

File: scripts/release.js [hunk 5]
```diff
@@ -20,29 +22,57 @@ const packageJson = require('../package.json');
...
+    });
   });
-});
+}
 
-function checkEnvironment() {
+function releaseStableVersion() {
+  shell.echo(`Algoliasearch-Helper release STABLE version`);
```
Comment (francoischalifour) [inline scripts/release.js]:
```suggestion
  shell.echo(`AlgoliaSearch Helper release STABLE version`);
```

File: scripts/release.js [hunk 6]
```diff
@@ -20,29 +22,57 @@ const packageJson = require('../package.json');
...
+  showChangelog(shell);
+  promptVersion(packageJson.version, (version) => {
+    bumpVersion(version, () => {
+      build();
+      updateChangelog(shell);
+      commitNewFiles(version);
+      publish();
+      goBackToDevelop();
+
+      shell.echo(`Algoliasearch-Helper v${version} released`);
```
Comment (francoischalifour) [inline scripts/release.js]:
```suggestion
      shell.echo(`AlgoliaSearch Helper v${version} released`);
```

File: scripts/release.js [hunk 7]
```diff
@@ -20,29 +22,57 @@ const packageJson = require('../package.json');
...
+
+  checkCleanWorkdir();
+  updateCanaryVersion(canaryVersion => {
     build();
-    updateChangelog(shell);
-    commitNewFiles(version);
-    publish();
-    goBackToDevelop();
+    publishOnNpm('canary');
+    revertStableVersion(packageJson.version, () => {
```
Comment (francoischalifour) [inline scripts/release.js]: Is this function completely resetting the state it was in before creating the canary release? (`version` in `package.json`, git staging, etc.)

It's quite hard to follow the release flow.

Comment (samouss) [reply scripts/release.js]: Yes, it does. I agree that it's not the optimal solution. A better approach would be to build the complete package into a dedicated folder that we can discard at the end. Unfortunately, that would completely change the current release process. I tried to minimize the changes, hence the update + revert of the changes.

File: scripts/release.js (hunk 3 above)
Comment (Haroenv) [reply scripts/release.js]: the package name is `algoliasearch-helper` though

File: scripts/release.js [hunk 8]
```diff
@@ -136,3 +170,16 @@ function goBackToDevelop() {
 function build() {
   shell.exec('yarn run build');
 }
+
+function updateCanaryVersion(cb) {
+  const lastCommitHash = shell.exec('git rev-parse --short HEAD', {silent: true}).toString().trim();
+  const canaryVersion = `0.0.0-${lastCommitHash}`;
```
Comment (Haroenv) [inline scripts/release.js]: as mentioned, this isn't incrementing, although I'm not too sure if that's an issue

File: scripts/release.js (hunk 7 above)
Comment (Haroenv) [reply scripts/release.js]: here it's still the old version, since you require before you change things right?

File: scripts/release.js (hunk 3 above)
Comment (francoischalifour) [reply scripts/release.js]: We should either go with the package name ("algoliasearch-helper") or the library title ("AlgoliaSearch Helper").

Comment (samouss) [reply scripts/release.js]: Yep, will use the name that we have on the Community website which is: Algolia JS Helper.

File: scripts/release.js (hunk 7 above)
Comment (samouss) [reply scripts/release.js]: Yes it is.

File: scripts/release.js (hunk 8 above)
Comment (samouss) [reply scripts/release.js]: I've tested on a package and it seems to work as expected. I manage to publish three versions on the `canary` tag with three different commit messages. I've checked that the hash does not follow each other in any way. It looks like it's working.

> https://yarnpkg.com/en/package/@samouss/useless

File: scripts/release.js (hunk 3 above)
Comment (samouss) [reply scripts/release.js]: https://github.com/algolia/algoliasearch-helper-js/pull/712/commits/6ebe9c7a44f33183c3cf04afc3285369663cebb6

File: scripts/release.js (hunk 5 above)
Comment (samouss) [reply scripts/release.js]: https://github.com/algolia/algoliasearch-helper-js/pull/712/commits/6ebe9c7a44f33183c3cf04afc3285369663cebb6

File: scripts/release.js (hunk 4 above)
Comment (samouss) [reply scripts/release.js]: https://github.com/algolia/algoliasearch-helper-js/pull/712/commits/6ebe9c7a44f33183c3cf04afc3285369663cebb6

File: scripts/release.js (hunk 6 above)
Comment (samouss) [reply scripts/release.js]: https://github.com/algolia/algoliasearch-helper-js/pull/712/commits/6ebe9c7a44f33183c3cf04afc3285369663cebb6

**Comment by francoischalifour** [comment]
Could we rely on a build number that is incremental? (or a hashed date)

The commit hash is great but there's no easy way to know if a build is more recent than another.

For example, [the dumb package released on unpkg](https://unpkg.com/@samouss/useless/) isn't listed based on the date. It's not easy to know which one is older.

**Comment by Haroenv** [comment]
on Yarn the list has a date though, I guess that's enough to work with, since usually you just want to use a specific version I guess?

**Comment by francoischalifour** [comment]
That's right. Having this info on the Yarn website is OK for me!
